### PR TITLE
Polish instructor revenue card (resolve #173)

### LINKED ISSUE #173: Polish revenue card component
We want to show instructors a visual view of their months of revenue. We currently display it in text form, but want to present it as a chart with the data that is available (currently views and revenue with month as the x axis and and amount as the y axis for the last 5 months).

The total amounts for the period should be listed in the new colors and sizes shown:

image

Instead of the orange estimated end, for now we can just list the latest month as another point.

This revenue is placed at the top of a published instructors dashboard:
image
https://app.zeplin.io/project.html#pid=58caceae349a0d1341e322a5&sid=58cba5bd34b78a2410327f41

## Implementation

- [x] We decided on using [this React Chart.js wrapper](http://gor181.github.io/react-chartjs-2/).
- [x] Upgrade text styles to latest mocks
- [x] Revenue chart points
- [x] Polish styles
- [x] Polish responsive scaling
- [x] Format axis numbers
- [x] Chart wrapper component
- [x] Add subscriber minutes overlay to see releationship of minutes to revenue over time
- [x] Add simple version to dashboard
- [x] Add detail route
- [ ] Repace hard-coded color values; relies on https://github.com/eggheadio/tachyons-egghead/issues/28

### PR DESCRIPTION
**Pull request by trevordmiller**
# Changes

- Polish revenue period text to match latest mocks
- Add react chart.js 2
- Add line charts for revenue and minutes for months available
- Abstract chart options and logic into `LineChart` component; could be a good candidate for promotion to egghead-ui later if we use it more
- Ensure everything should be flexible if endpoints change months returned
- Promote `InstructorRevenueCard` up and render it on `Instructor` screen as well as `Dashboard`, if the viewing user has `instructor.revenue_url` (so Joel can view everyone's revenue, and instructors can see their own revenue on their individual instructor page as well as their dashboard)

# Result For User

revenue-charts

# Note

@vojtaholik I tried to match the mocks, but I decided to move the text from the left to always being on top (stacked or horizontal if large enough screen) because the graphs render better with more horizontal space. I also had to split the graphs into two, stacked on top of each other, because when they were combined the scale was much too large for most instructors and made your revenue seem very small:
screen shot 2017-03-30 at 3 41 19 pm

You can modify the relative scale that chartjs uses, but because minutes will always be much higher than revenue, I'm not sure that that makes sense to have them overlap.

---

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -16,14 +16,17 @@
     "react-scripts": "^0.8.4"
   },
   "dependencies": {
+    "chart.js": "^2.5.0",
```
Comment (trevordmiller) [inline package.json:19]: Required as a separate dep from `react-chartjs-2` per:
image

File: src/components/InstructorRevenueCard/components/LineChart/index.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
+import React from 'react'
+import {map} from 'lodash'
+import hexToRgba from 'hex-rgba'
+import {Line} from 'react-chartjs-2'
+import numberFormattingByType from 'utils/numberFormattingByType'
+
+const sharedDataWithColor = (color) => ({
+  borderColor: color,
+  backgroundColor: hexToRgba(color, 20),
```
Comment (trevordmiller) [inline src/components/InstructorRevenueCard/components/LineChart/index.js]: Auto-calculate background fill as a light shade of main color:
image
image

File: src/components/InstructorRevenueCard/components/LineChart/index.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
+import React from 'react'
+import {map} from 'lodash'
+import hexToRgba from 'hex-rgba'
+import {Line} from 'react-chartjs-2'
+import numberFormattingByType from 'utils/numberFormattingByType'
```
Comment (trevordmiller) [inline src/components/InstructorRevenueCard/components/LineChart/index.js:5]: I promoted this logic to be used in multiple places

File: src/components/InstructorRevenueCard/components/LineChart/index.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+          }
+        }],
+      },
+      tooltips: {
+        enabled: true,
+        mode: 'single',
+        callbacks: {
+          label: (tooltipItems) => (
+            numberFormattingByType.general(tooltipItems.yLabel)
+          ),
```
Comment (trevordmiller) [inline src/components/InstructorRevenueCard/components/LineChart/index.js]: Adds commas to numbers larger than `999`

File: src/components/InstructorRevenueCard/components/LineChart/index.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        },
+      },
+    }}
+    data={{
+      labels: xAxis,
+      datasets: map(yAxis, set => ({
+        ...sharedDataWithColor(set.color),
```
Comment (trevordmiller) [inline src/components/InstructorRevenueCard/components/LineChart/index.js]: Abstract color

File: src/components/InstructorRevenueCard/index.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+      {({data}) => {
+        const currentMonthRevenue = find(data, ['month', currentMonthStartDate()])
+        const currentTotalRevenue = totalRevenue(removeRevenueMonth(data, currentMonthStartDate()))
+        const hasCurrentRevenue = size(data) > 0
+          && ((currentMonthRevenue && currentMonthRevenue.revenue > 0) || (currentTotalRevenue && currentTotalRevenue.revenue > 0))
+        const currentMonthDates = map(data, month => month.month)
+        const currentRevenuePoints = map(data, month => month.revenue)
+        const currentMinutesPoints = map(data, month => month.minutes_watched)
+        const revenueColor = '#4786ff' // # blue
+        const minutesColor = '#171e27' // # base-secondary
```
Comment (trevordmiller) [inline src/components/InstructorRevenueCard/index.js:29]: Could remove these hard coded hex if this tachyons story is implemented:
https://github.com/eggheadio/tachyons-egghead/issues/28

File: src/components/InstructorRevenueCard/index.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+                  <LineChart
+                    xAxis={currentMonthDates}
+                    yAxis={[
+                      {
+                        color: minutesColor,
+                        points: currentMinutesPoints,
+                      },
+                    ]}
+                  />
+                </div>
```
Comment (trevordmiller) [inline src/components/InstructorRevenueCard/index.js]: This is the public interface for the customized parts of the line charts now, after all the abstracting in `LineChart` of chartjs config

**Review by joelhooks — APPROVED** [review]
nice to see some charts! 

I'm not stoked on all the labels and huge amount of real estate that they occupy. We can scale the "minutes watched" to hours or some other increment that would make them combine better in the overlaid original design. If that isn't doing it, we can drop minutes and just dhow revenue for now.

File: src/components/LayoutColumns/index.js [hunk 8]
```diff
@@ -7,6 +7,7 @@ export default ({items, relativeSizes}) => (
       <div
         key={index}
         className={`
+          mb4
```
Comment (trevordmiller) [inline src/components/LayoutColumns/index.js:10]: Layout columns take care of gutters by device size (horizontal and vertical) so components don't need wrapping padding/margin

File: src/screens/Instructors/screens/Instructor/index.js [hunk 9]
```diff
@@ -10,14 +11,20 @@ export default ({instructor}) => (
...
+    <LayoutColumns
+      items={[
+        <InstructorStatsCard
+          publishedLessons={instructor.published_lessons}
+          publishedCourses={instructor.published_courses}
+        />,
+        <InstructorRevenueCard revenueUrl={instructor.revenue_url} />,
+      ]}
+      relativeSizes={[1, 2]}
+    />
```
Comment (trevordmiller) [inline src/screens/Instructors/screens/Instructor/index.js:28]: Adds revenue to `Instructor` screen if you have access to view it:
image

**Comment by trevordmiller** [comment]
> "I'm not stoked on all the labels and huge amount of real estate that they occupy. We can scale the "minutes watched" to hours or some other increment that would make them combine better in the overlaid original design. If that isn't doing it, we can drop minutes and just dhow revenue for now."

@joelhooks @vojtaholik - only rendering the revenue chart makes the most sense to me right now. What do you think Vojta? This is what it would look like:

image
image

Although for the future, I do think hours would be a more human-friendly unit than minutes in general for the endpoints and daily/monthly report emails.

**Comment by trevordmiller** [comment]
Another thought is I could easily replace these dates to just be the month names:
image

This would work well for now since we only go back 5 months.

**Comment by joelhooks** [comment]
Dropping the labels isn't likely to leave anybody terribly confused, and should generally be more flexible with the dynamic time range. It'd also be closer to the comp, which is good too.

**Comment by trevordmiller** [comment]
> "Dropping the labels isn't likely to leave anybody terribly confused"

@joelhooks You mean drop the x-axis labels completely? I think having the month name at least would be valuable and take up less space than the full date; would look like this:

image

image

But if you and Vojta would prefer no x-axis labels then we can do that.

**Comment by trevordmiller** [comment]
^ That would also allow the desktop rendering to better match the mocks with the horizontal layout:
image

It just depends if you want the chart to be more fat or tall.

**Review by tayiorbeii — APPROVED** [review]

**Comment by joelhooks** [comment]
image

I like this quite a bit. I like how it fills the viewport and uses linear non bezier points and I also prefer the text spacing with no wraps. Can it be more like the comp in those ways?

**Comment by trevordmiller** [comment]
@joelhooks Yes, the container can have more space alloted, but will still wrap if the screen size is small enough. As for removing the labels and points, how do I know what the data is? Like, how many minutes and revenue I had in January?

**Comment by trevordmiller** [comment]
To have the overlapping charts, I don't understand why the taller one is green; I assumed that would be the minutes watched (which I could convert to hours for a smaller scale), but it appears to be attached to the latest month, which wouldn't fit the same scale as the last five months.

**Comment by joelhooks** [comment]
We can skip the overlapping, but this is a visual flourish. We will have a more detailed breakdown of stats on another screen in the future. This is an "at a glance heartbeat". Revenue and views will mostly look the same graphed (with a scaled y) so a single non-overlapping graph is adequate.

**Comment by vojtaholik** [comment]
I agree we can skip overlapping. 
If we are going to display X axis with months I think it'd be great to show a revenue for each month as well. I'd be interested in such info.

Mockup:
image




What about showing current month as a green line at the end and hover would show "This month so far" $? This way it could be merged into one chart.

So basically a timeline with last 5 months revenue and current month in different color.

**Comment by joelhooks** [comment]
That looks better, but how will it look with the labels if we changed it to 12 months, for instance. 

I don't know if the charting library can get fancy like that with the current month.

**Comment by trevordmiller** [comment]
@joelhooks @vojtaholik After reading through the comments on this issue again, I think what I'll do is:

1) Update the dashboard chart to look like:
image
No labels / points / curves / interactions, and full width/height.

2) Since I already have the code in place for a more detailed view, I can add a new `/revenue` route that renders a larger, more detailed chart when you tap on the simple chart from #1 ^ in the dashboard. It would be very easy for me to add this (will just add a `detail` option to the `LineChart` wrapper I made) and then we can iterate on it later; but that way the minimal non-interactive version on the dashboard when tapped could at least take you to this more detailed version so you can see specific points / values.

Does that sound good?

File: src/components/OpenToggle/index.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+  render() {
+    const {isOpen} = this.state
+    const {children} = this.props
+    return children({
+      isOpen,
+      handleOpenToggleClick: this.handleOpenToggleClick,
+    })
+  }
+}
```
Comment (trevordmiller) [inline src/components/OpenToggle/index.js:24]: I made this new Render Callback component (aka function as child), now that this logic is being shared across more than one component.

File: src/components/RequestedLessonsCard/components/ProposeLesson/index.js [hunk 11]
```diff
@@ -1,27 +1,15 @@
...
-    return isOpen
+export default ({instructor}) => (
+  <OpenToggle>
+    {({isOpen, handleOpenToggleClick}) => isOpen 
       ? <ProposeLessonForm instructor={instructor} />
       : <div 
-          onClick={this.handleClick}
+          onClick={handleOpenToggleClick}
           className='ttu'
         >
```
Comment (trevordmiller) [inline src/components/RequestedLessonsCard/components/ProposeLesson/index.js:14]: Another example of how the render callback is used

File: src/components/InstructorRevenueCard/index.js [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+
+        const currentMonthDates = map(data, month => month.month)
+        const currentRevenuePoints = map(data, month => month.revenue)
+        const currentMinutesPoints = map(data, month => month.minutes_watched)
+        const revenueColor = '#4786ff' // # blue
+        const minutesColor = '#171e27' // # base-secondary
+
+        return (
+          <OpenToggle>
+            {({isOpen, handleOpenToggleClick}) => (
```
Comment (trevordmiller) [inline src/components/InstructorRevenueCard/index.js:33]: Using the new render callback

File: src/components/InstructorRevenueCard/index.js [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+                      xAxis={currentMonthDates}
+                      yAxis={[
+                        {
+                          color: revenueColor,
+                          points: currentRevenuePoints,
+                        },
+                      ]}
+                      className='bg-white-secondary br2 br--bottom'
+                    />
+                  }
```
Comment (trevordmiller) [inline src/components/InstructorRevenueCard/index.js:114]: See three different `<LineChart` instances above; this is the abstraction I made over the ChartJS which requires a lot more config. So now you just need to pass the axis's and `type` (defaults to `simple` but can be `detail`).
Could be a good candidate for promotion to egghead-ui, per the work @tayiorbeii has been doing with chartjs.

**Comment by trevordmiller** [comment]
@joelhooks @vojtaholik @tayiorbeii @ijones16 

I updated this pull request with some more commits and inline comments. Revenue now better matches the mocks by default, but has a button to view details, which render in place:
revenue-details

I'm going to merge this now, but am open to feedback for future iterations.

**Comment by joelhooks** [comment]
Looks great when it loads, but I'm not a fan of the new large orange button.

If that detail view is going to exist now, can just tapping on the chart "flip" the card and show the detailed view? This is instead of a button. 

Had some questions about the detail variant:

Can we simplify the labels similar to what @vojtaholik proposed on the detail view, and also fill the available space? https://github.com/eggheadio/egghead-instructor-center/pull/207#issuecomment-290674989

It shifts the UI row below it down visually, can it be the same size as the simple view to prevent that?

Can we add some sort of minor transition between views?

If I'm sick of looking at details and want to see the previous view, will I need to reload the page?

**Comment by trevordmiller** [comment]
@joelhooks I agree with your comments. Moved with details to another story: https://github.com/eggheadio/egghead-instructor-center/issues/209

**Pull request merged**